File: src/pr_reviewer/events/record_event.py
"""Hosted lifecycle events (Runtime Task 1B).

A hosted event row is a redacted lifecycle event: identifiers, enums, and aggregate numbers,
never a nested structure. serialize_json_object enforces that here, at the application layer, by
rejecting any payload value that is itself an object or array -- not a validated string, a flat
shape with no depth for a findings list, a diff, or a sandbox log to hide in. migration
202608291930_rescope_hosted_events.sql's agent_events_payload_is_flat CHECK constraint enforces
the identical rule at the database layer, so a writer that reaches the table through raw SQL
rather than this function is still stopped.

JsonObject/JsonValue stay fully recursive: they describe whatever a jsonb column can hold in
general (this file's read path, and other jsonb columns elsewhere), not the narrower flat shape
this file's write path now enforces.
"""

from __future__ import annotations

import json
import math
from collections.abc import Mapping
from dataclasses import dataclass
from datetime import datetime

from pr_reviewer.db.client import connection

type JsonPrimitive = bool | int | float | str | None
type JsonValue = JsonPrimitive | list[JsonValue] | dict[str, JsonValue]
type JsonObject = dict[str, JsonValue]


@dataclass(frozen=True)
class AgentEvent:
    id: str
    sequence: int
    review_job_id: str | None
    event_type: str
    payload: JsonObject
    created_at: datetime


def record_event(review_job_id: str, event_type: str, payload: JsonObject) -> None:
    with connection() as conn:
        conn.execute(
            """
            insert into agent_events (review_job_id, event_type, payload)
            values (%s, %s, %s::jsonb)
            """,
            (review_job_id, event_type, serialize_json_object(payload)),
        )


def serialize_json_object(value: JsonObject) -> str:
    assert_flat_json_object(value)
    return json.dumps(value, separators=(",", ":"))


def assert_flat_json_object(value: JsonObject) -> None:
    """The top-level object's values must be JSON scalars: no nested object, no nested array.
    A cyclic structure is necessarily a nested object (a dict cannot contain itself as a scalar),
    so rejecting nesting outright also rejects a cycle, with no need to track visited ids.
    """
    if not isinstance(value, Mapping):
        raise TypeError("Expected a JSON object for a hosted event payload")
    for key, item in value.items():
        if not isinstance(key, str):
            raise TypeError(f"Expected a string key, got {key!r}")
        assert_json_scalar(item, key)


def assert_json_scalar(value: object, key: str) -> None:
    if value is None or isinstance(value, bool | str):
        return

    if isinstance(value, int):
        return

    if isinstance(value, float):
        if not math.isfinite(value):
            raise TypeError(f"Expected a finite JSON number at {key!r}")
        return

    raise TypeError(
        f"Expected a flat JSON scalar (bool, int, float, str, or None) at {key!r}; a hosted "
        "event payload cannot carry a nested object or array"
    )

File: packages/core/src/events/recordEvent.ts
import { db } from "../db/client";

export type JsonPrimitive = boolean | number | string | null;
export type JsonValue = JsonPrimitive | JsonObject | JsonValue[];
export type JsonObject = { [key: string]: JsonValue };

export type KnownAgentEventType =
  | "webhook.accepted"
  | "model_call.recorded"
  | "review_job_failed"
  | "review_job_retry_scheduled";

export type AgentEventType = KnownAgentEventType | (string & {});

export type RecordEventInput = {
  reviewJobId: string;
  eventType: AgentEventType;
  payload: JsonObject;
};

export type AgentEvent = {
  id: string;
  sequence: string;
  reviewJobId: string | null;
  eventType: AgentEventType;
  payload: JsonObject;
  createdAt: Date;
};

export async function recordEvent(input: RecordEventInput): Promise<void> {
  await db.query(
    `insert into agent_events (review_job_id, event_type, payload)
     values ($1, $2, $3::jsonb)`,
    [input.reviewJobId, input.eventType, serializeJsonObject(input.payload)],
  );
}

export function serializeJsonObject(value: JsonObject): string {
  assertJsonValue(value, "$", new Set<object>());
  return JSON.stringify(value);
}

function assertJsonValue(value: unknown, path: string, ancestors: Set<object>): asserts value is JsonValue {
  if (value === null || typeof value === "boolean" || typeof value === "string") {
    return;
  }

  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new TypeError(`Expected a finite JSON number at ${path}`);
    }
    return;
  }

  if (Array.isArray(value)) {
    assertAcyclic(value, path, ancestors);
    for (const [index, item] of value.entries()) {
      assertJsonValue(item, `${path}[${index}]`, ancestors);
    }
    ancestors.delete(value);
    return;
  }

  if (typeof value === "object") {
    const prototype = Object.getPrototypeOf(value);
    if (prototype !== Object.prototype && prototype !== null) {
      throw new TypeError(`Expected a JSON object at ${path}`);
    }

    assertAcyclic(value, path, ancestors);
    for (const [key, item] of Object.entries(value)) {
      assertJsonValue(item, `${path}.${key}`, ancestors);
    }
    ancestors.delete(value);
    return;
  }

  throw new TypeError(`Expected a JSON value at ${path}`);
}

function assertAcyclic(value: object, path: string, ancestors: Set<object>): void {
  if (ancestors.has(value)) {
    throw new TypeError(`Expected an acyclic JSON value at ${path}`);
  }
  ancestors.add(value);
}
